src/lib/storage/reference_column/reference_column_iterable.hpp

```diff
@@ -17,74 +15,96 @@ class ReferenceColumnIterable : public ColumnIterable<ReferenceColumnIterable<T>
 
   template <typename Functor>
   void _on_with_iterators(const Functor& functor) const {
-    const auto table = _column.referenced_table();
-    const auto column_id = _column.referenced_column_id();
+    const auto pos_list = _column.pos_list();
+
+    const auto chunk_id_and_chunk_offsets_list = to_chunk_id_and_chunk_offsets_list(*pos_list);
+    if (chunk_id_and_chunk_offsets_list) {
+      const auto& [referenced_chunk_id, chunk_offsets_list] = *chunk_id_and_chunk_offsets_list;
+
+      _referenced_with_iterators(referenced_chunk_id, chunk_offsets_list, functor);
+      return;
+    }
 
-    const auto begin_it = _column.pos_list()->begin();
-    const auto end_it = _column.pos_list()->end();
+    const auto column_id = _column.referenced_column_id();
+    const auto table = _column.referenced_table();
 
-    auto begin = Iterator{table, column_id, begin_it, begin_it};
-    auto end = Iterator{table, column_id, begin_it, end_it};
+    auto begin = Iterator{column_id, table.get(), pos_list.get(), ChunkOffset{0u}};
+    auto end = Iterator{static_cast<ChunkOffset>(pos_list->size())};
     functor(begin, end);
   }
 
  private:
   const ReferenceColumn& _column;
 
  private:
+  template <typename Functor>
+  void _referenced_with_iterators(const ChunkID chunk_id, const ChunkOffsetsList& chunk_offsets_list,
+                                  const Functor& functor) const {
+    const auto column_id = _column.referenced_column_id();
+    const auto table = _column.referenced_table();
+
+    const auto base_column = table->get_chunk(chunk_id)->get_column(column_id);
+
+    resolve_column_type<T>(*base_column, [&functor, &chunk_offsets_list](const auto& referenced_column) {
+      using ColumnT = std::decay_t<decltype(referenced_column)>;
+      constexpr auto is_reference_column = std::is_same_v<ColumnT, ReferenceColumn>;
+
+      // clang-format off
+      if constexpr (!is_reference_column) {
+        auto iterable = create_iterable_from_column<T>(referenced_column);
+        iterable.with_iterators(&chunk_offsets_list, [&](auto it, auto end) { functor(it, end); });
+      } else {
+        Fail("Reference column must not reference another reference column.");
```

Comment:
Guard Clause?

Reply:
Doesn’t work with `if constexpr` I think.

Reply:
static_assert? :D

Reply:
Theoretically, this is never going to happen. I could replace the `else` block with something like `DebugAssert(is_reference_column, "Error message");`, but would this make it better?